### ISSUE TITLE
A top corner pixel is missing in non-filled `CanvasItem.draw_rect()`

### ISSUE THREAD

**Issue by fire-forge** [godotengine/godot#58775]
### Godot version

v4.0.alpha.custom_build [013efeef0]

### System information

Windows 10, Intel Core i5 7200U CPU, Vulkan

### Issue description

When using `CanvasItem.draw_rect()` with `fill = false` and `width = 1`, one of the top corner pixels are missing. From my observations, the top-left corner is missing when the pixel position is an integer and the top-right corner is missing when the pixel position has a decimal.

This also affects some areas of the editor, such as the 2D rectangle selection and the gradient editor:
![image](https://user-images.githubusercontent.com/67974470/156868084-017e2d52-2a9a-4737-b3b7-24e5b358b52e.png) ![image](https://user-images.githubusercontent.com/67974470/156868307-6f6174a3-27df-4211-b5c8-9ea1054b3a9e.png)

### Steps to reproduce

Each of these 3 options exhibits this bug:
a. Open the 2D editor and start a rectangle selection
b. Create a Gradient resource and add some points to the gradient editor
c. Attach a script to a CanvasItem-inherited node, and use `draw_rect()` with `fill = false` in the `_draw()` function.

**Comment by Calinou** [comment]
@fire-forge Can you reproduce this in 3.4.3?

**Comment by fire-forge** [comment]
@Calinou in 3.4.3, the bottom-left corner is missing with an integer position and nothing is missing with a decimal position. 

![image](https://user-images.githubusercontent.com/67974470/156896398-a4e3b34b-3376-483b-a040-50c8b23d23b1.png)

Integer position:
![image](https://user-images.githubusercontent.com/67974470/156896454-112fd99a-7a75-49f8-95b2-22d0bfd35a55.png)

Decimal position:
![image](https://user-images.githubusercontent.com/67974470/156896448-4d96f1ce-8e90-4d6f-a264-be9ee8ffdf3e.png)

**Comment by mrimvo** [comment]
I think I got the same issue in Godot 4.0 Beta 4.

![image](https://user-images.githubusercontent.com/6030346/202667077-fd2996c8-e1a5-4f7b-8447-716e158f15e5.png)

The corner pieces of the rectangle are missing, iff `rect.x<0`.

```draw_rect(rect, Color.LIME_GREEN, false, 6)```

**Comment by MewPurPur** [comment]
@mrimvo This looks like a different bug, where negative X results in too short lines, and negative Y results in too long lines. I know how to fix it, but want to first see if daleexev's PR will fix it.

**Comment by dalexeev** [comment]
Quoting kleonc's [comment](https://github.com/godotengine/godot/pull/41239#issuecomment-1384421182):

> I think #58775 is not really solvable in a general case. It's not just a matter of obtaining proper final positions of the line endpoints in the coordinate space of the viewport/framebuffer/render target (or whatever you call it) which would be needed to be able to apply a proper offset in that space (small final offset; regardless of the zoom etc.). What we'd also need to know to determine what such offset should be is the exact algorithm used by the Vulkan/OpenGL/etc. for rasterizing the line primitives. And that's problematic, because these can differ between implementations since specifications aren't super strict:
> 
>     * [OpenGL 3.3 Basic Line Segment Rasterization](https://registry.khronos.org/OpenGL/specs/gl/glspec33.core.pdf#subsection.3.5.1) (for [4.6 it's pretty much the same](https://registry.khronos.org/OpenGL/specs/gl/glspec46.core.pdf#subsection.14.5.1)),
> 
>     * [Vulkan 1.1 Line Segments Rasterization](https://registry.khronos.org/vulkan/specs/1.1/html/chap25.html#primsrast-lines).
> 
> 
> So even assuming that rasterization is done according to the diamond exit rule (which is not guaranteed; but maybe it's actually the most common case?thinking no idea about that) we would still need to find a way to somehow obtain the final positions of the line endpoints within the fragments' coordinate space (I don't see how we could currently do it). And then we would still need to find a way of how to offset the vertices (probably branching into many different cases) to always produce satisfying fragments. E.g. take a look at the image [from this post](https://stackoverflow.com/a/10341863) and imagine all possible configurations of the vertices (colored dots); and yeah, the rect can end up being transformed in there (so take into consideration rotations or... skewing; yeah, good luck with that smile).
> 
> **TL;DR**: I don't see us solving it. Would love to be surprised by someone but I don't think it will happen.
> 
> For a non-general solution / hacky solution what maybe could be done is adding a method (are a bool param to the current `draw_rect`) which would be supposed to work properly only in a very specific conditions: when the given CanvasItem is not transformed so the offset we would apply to the vertices would actually be equivalent to the offset in the fragments' space. Still it wouldn't be guaranteed to always work because of the loose specifications mentioned previously. Anyway, I'm not convinced that trying to handle such edge case is a good idea.

A few hints:

1. #69851 fixed this for positive `width`s, this issue only applies to negative `width`s. If you explicitly pass `width = 1`, then the corners should not be missing or brighter.
2. Thin rectangles can indeed be rendered slightly incorrectly, but it won't be very noticeable as lines with negative `width`s remain thin when the `CanvasItem` is scaled. As long as you don't stretch the `Viewport` image. If you expect this (for example in pixel games), then you should always pass `width = 1`.
3. See godotengine/godot-docs#6455. This is not obvious, but important information that will allow you to avoid some problems.

**Comment by Calinou** [comment]
I can confirm this on 4.5.beta 4d1f26e1f on all renderers (Linux + NVIDIA).

**Testing project:** [test_draw_rect_corner.zip](https://github.com/user-attachments/files/21172439/test_draw_rect_corner.zip)

https://github.com/godotengine/godot/pull/86465 helps with this for rects that are offset by 0.5 pixels or more (e.g. `Rect2(10.5, 10.5, 60, 40)`).

*From top to bottom: No offset, 0.25 pixel offset, 0.5 pixel offset, 0.75 pixel offset (applied on both axes). Rects on the left use a width of `-1.0` (default), while rects on the right use a width of `1.0` (1 pixel-thick polygons).*

MSAA | `master` | Bresenham PR
-|-|-
Off | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/217a62fe-065b-42ee-9508-a997143f9c8f" /> | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/ac39b4dd-b697-442a-a415-f5f9a77b628d" />
2× | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/180f3f21-b043-4b87-b575-b7aafa7b8271" /> | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/0c239d8a-8856-4a41-b615-466f0dfa3f3d" />
4× | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/0de7c6e4-81ad-4061-86fc-764e4e8539f1" /> | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/b9694d2a-98a9-452e-a5f8-a04d740e5e85" />
8× | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/8eeb0db1-38bd-4d89-addb-2e3f16c39199" /> | <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/b40ff3de-ff0d-46aa-b184-495484ee8e4f" />

The viewport is 128×128, images are scaled to 768×768 when viewed at full size.

**Comment by samw-26** [comment]
Adding this in case anyone needs it. The workaround for setting width to a positive value will not work for different zoom levels, you need to divide by the scale of the viewport transform.
```gdscript
draw_rect(Rect2(start_position, end_position), Color.GREEN, false, 1.0 / get_viewport_transform().get_scale().x)
```
